src/esp/assets/GenericInstanceMeshData.cpp

```diff
@@ -94,33 +105,184 @@ Cr::Containers::Optional<InstancePlyData> parsePly(
+      std::size_t numSSDObjs = ssdObjs.size();
+
+      // map of colors as ints as key, where value is objectID from ssd
+      std::unordered_map<uint32_t, int> tmpColorMapToSSDid;
+      // map of colors as ints as key, where value is regionID from ssd
+      std::unordered_map<uint32_t, int> tmpColorMapToRegionIndex;
+      // lambda function to convert a 3-vec representation of a color into an
+      // int
+      auto colorAsInt = [](const Mn::Vector3ub& color) -> uint32_t {
+        return (((color[0] * 256) + color[1]) * 256) + color[2];
+      };
+
+      // build maps of color ints to semantic IDs and color ints to region ids
+      // in SSD
+      // find max regions present, so we can have an "unassigned"/"unknown"
+      // region to move all verts whose region is
+
+      int maxRegion = -1;
+      for (int i = 0; i < numSSDObjs; ++i) {
+        const auto& ssdObj =
+            static_cast<scene::HM3DObjectInstance&>(*ssdObjs[i]);
+        const uint32_t colorInt = colorAsInt(ssdObj.getColor());
+        int regionIDX =
+            static_cast<scene::HM3DSemanticRegion&>(*ssdObj.region())
+                .getIndex();
+        tmpColorMapToSSDid[colorInt] = ssdObj.getSemanticID();
+        tmpColorMapToRegionIndex[colorInt] = regionIDX;
+        maxRegion = (regionIDX > maxRegion ? regionIDX : maxRegion);
+      }
+
+      // increment maxRegion to use for unknown regions whose colors are not
+      // mapped
+      ++maxRegion;
+
+      // rebuild objectIDs vector (objectID per vertex) based on
+      // per vertex colors mapped
+      // 1st semantic ID for colors not found in SSD
+      std::size_t nonSSDObjID = tmpColorMapToSSDid.size();
+      // color map with first nonSSDObjID-1 elements in proper order to match
+      // provided semantic IDs (so that ID is IDX of semantic color in map) and
+      // any overflow colors are use an unmapped semantic ID as their index.
+      Cr::Containers::Array<Mn::Vector3ub> tmpDestSSDColorMap{
+          Mn::NoInit, tmpMeshColorMap.size()};
+
+      // map regionIDs to affected verts using semantic color provided in
+      // SemanticScene, as specified in SSD file.
+      data.partitionIds.resize(meshData->vertexCount());
+      // map semantic IDs corresponding to colors from SSD file to appropriate
+      // verts (via index)
+      data.objectIds.resize(meshData->vertexCount());
+      // temporary holding structure to hold any non-SSD vert colors, so that
+      // the nonSSDObjID for new colors can be incremented appropriately
+      // not using set to avoid extra include
+      std::unordered_map<uint32_t, int> nonSSDVertColors;
+
+      // only go through all verts one time
+      for (int vertIdx = 0; vertIdx < tmpMeshObjectIds.size(); ++vertIdx) {
+        // Get mesh-derived objectID @ vert
+        const int objIdAtVert = tmpMeshObjectIds[vertIdx];
+        // Get meshcolor @ vertexobjPartitionsFromSSD
+        Mn::Vector3ub meshColor = tmpMeshColorMap[objIdAtVert];
```

Comment:
I wanted to move this code to after this block too, but I need the number of total unique colors in the mesh in this block of code so that I could reserve the size of the `tmpDestSSDColorMap`, since I need to be able to place a particular vertex's color within `tmpDestSSDColorMap` at the index defined by the semantic ID, either from the ssd map, or from the "unassigned" region at the end of the vertex list (for colors not in the ssd).    

Instead of performing the entire index and color mapping, beforehand, though, I can just build a map of id to color first, and then copy from -that- map into the appropriate pre-allocated vector.  Still -much- cheaper than going through the verts before-hand.